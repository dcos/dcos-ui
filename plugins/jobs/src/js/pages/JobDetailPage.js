import { i18nMark } from "@lingui/react";

/* eslint-disable no-unused-vars */
import React from "react";
/* eslint-enable no-unused-vars */
import { routerShape } from "react-router";
import PropTypes from "prop-types";

import mixin from "reactjs-mixin";

import Page from "#SRC/js/components/Page";
import TabsMixin from "#SRC/js/mixins/TabsMixin";
import Job from "#SRC/js/structs/Job";
import Util from "#SRC/js/utils/Util";

import JobFormModal from "../components/JobsFormModal";
import JobConfiguration from "./JobConfiguration";
import { DIALOGS } from "../JobDetailPageContainer";
import JobRunHistoryTable from "./JobRunHistoryTable";
import ItemSchedule from "../components/breadcrumbs/Schedule";
import ItemStatus from "../components/breadcrumbs/Status";
import Breadcrumbs from "../components/Breadcrumbs";
import JobDelete from "../JobDelete";

import jobsMenu from "../jobsMenu";

class JobDetailPage extends mixin(TabsMixin) {
  constructor() {
    super(...arguments);
    this.renderBreadcrumbStates = this.renderBreadcrumbStates.bind(this);

    this.tabs_tabs = {
      runHistory: i18nMark("Run History"),
      configuration: i18nMark("Configuration")
    };

    this.state = {
      currentTab: Object.keys(this.tabs_tabs).shift()
    };
  }

  getNavigationTabs() {
    return <ul className="menu-tabbed">{this.tabs_getUnroutedTabs()}</ul>;
  }

  renderConfigurationTabView(job) {
    return <JobConfiguration job={job} />;
  }

  renderRunHistoryTabView(job) {
    return <JobRunHistoryTable job={job} />;
  }

  renderBreadcrumbStates(item) {
    const status = Util.findNestedPropertyInObject(
      item,
      "jobRuns.longestRunningActiveRun.tasks.longestRunningTask.status"
    );

    let schedule = null;
    if (
      item.schedules &&
      item.schedules.nodes.length &&
      item.schedules.nodes[0].enabled
    ) {
      schedule = item.schedules.nodes[0];
    }

    return [
      schedule ? <ItemSchedule key="schedule" schedule={schedule} /> : null,
      status ? <ItemStatus key="status" status={status} /> : null
    ];
  }

  getActions() {
    const job = this.props.job;

    const customActionHandlers = {
      edit: this.props.handleEditButtonClick,
      delete: this.props.handleDestroyButtonClick
    };

    return jobsMenu(job, customActionHandlers);
  }

  getTabs() {
    const activeTab = this.state.currentTab;

    return [
      {
        label: i18nMark("Run History"),
        callback: () => {
          this.setState({ currentTab: "runHistory" });
        },
        isActive: activeTab === "runHistory"
      },
      {
        label: i18nMark("Configuration"),
        callback: () => {
          this.setState({ currentTab: "configuration" });
        },
        isActive: activeTab === "configuration"
      }
    ];
  }

  render() {
    if (this.props.params.taskID) {
      return this.props.children;
    }

    const { job, params } = this.props;

    return (
      <Page>
        <Page.Header
          actions={this.getActions()}
          breadcrumbs={
            <Breadcrumbs
              jobPath={job.path}
              jobName={job.name}
              jobInfo={this.renderBreadcrumbStates(job)}
            />
          }
          tabs={this.getTabs()}
        />
        {this.tabs_getTabView(job)}
        <JobFormModal
          isEdit={true}
          job={new Job(JSON.parse(job.json))}
          isOpen={this.props.jobActionDialog === DIALOGS.EDIT}
          closeModal={this.props.closeDialog}
        />
        <JobDelete
          jobId={params.id}
          open={this.props.jobActionDialog === DIALOGS.DESTROY}
          onSuccess={this.props.onJobDeleteSuccess}
          onClose={this.props.closeDialog}
        />
      </Page>
    );
  }
}

JobDetailPage.contextTypes = {
  router: routerShape
};

JobDetailPage.propTypes = {
  children: PropTypes.any,
  closeDialog: PropTypes.func,
  handleEditButtonClick: PropTypes.func.isRequired,
  handleDestroyButtonClick: PropTypes.func.isRequired,
  onJobDeleteSuccess: PropTypes.func.isRequired,
  job: PropTypes.shape({
    json: PropTypes.string.isRequired
  }),
  jobActionDialog: PropTypes.any
};

module.exports = JobDetailPage;
