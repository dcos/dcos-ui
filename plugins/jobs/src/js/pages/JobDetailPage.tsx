import { i18nMark } from "@lingui/react";
import * as React from "react";
import { routerShape } from "react-router";
import Page from "#SRC/js/components/Page";
import JobFormModal from "../components/JobsFormModal";
import JobConfiguration from "./JobConfiguration";
import JobRunHistoryTable from "./JobRunHistoryTable";
import ItemSchedule from "../components/breadcrumbs/Schedule";
import ItemStatus from "../components/breadcrumbs/Status";
import Breadcrumbs from "../components/Breadcrumbs";
import JobDelete from "../JobDelete";
import jobsMenu from "../jobsMenu";

class JobDetailPage extends React.Component<
  {
    job: {
      id: string;
      path?: string[];
      name?: string;
      json: string;
    };
    params: {
      id: string;
      taskID: string;
    };
  },
  { currentTab: string; currentDialog: null | "edit" | "destroy" }
> {
  static contextTypes = { router: routerShape };
  state = {
    currentDialog: null,
    currentTab: "runHistory",
  };

  onJobDeleteSuccess() {
    this.closeDialog();
    this.context.router.push("/jobs/overview");
  }

  renderBreadcrumbStates(item) {
    const schedule = item.schedules?.nodes?.find((n) => n.enabled);
    const status =
      item?.jobRuns?.longestRunningTask?.tasks?.longestRunningTask?.status;

    return [
      schedule ? <ItemSchedule key="schedule" schedule={schedule} /> : null,
      status ? <ItemStatus key="status" status={status} /> : null,
    ];
  }

  closeDialog() {
    this.setState({ currentDialog: null });
  }

  render() {
    if (this.props.params.taskID) {
      return this.props.children;
    }

    const jobDeleteSuccess = () => this.onJobDeleteSuccess();
    const closeDialog = () => this.closeDialog();
    const { job, params } = this.props;
    const setTab = (currentTab: string) => () => {
      this.setState({ currentTab });
    };

    return (
      <Page>
        <Page.Header
          actions={jobsMenu(this.props.job, {
            edit: () => void this.setState({ currentDialog: "edit" }),
            delete: () => void this.setState({ currentDialog: "destroy" }),
          })}
          breadcrumbs={
            <Breadcrumbs
              jobPath={job.path}
              jobName={job.name}
              jobInfo={this.renderBreadcrumbStates(job)}
            />
          }
          tabs={[
            {
              label: i18nMark("Run History"),
              callback: setTab("runHistory"),
              isActive: this.state.currentTab === "runHistory",
            },
            {
              label: i18nMark("Configuration"),
              callback: setTab("configuration"),
              isActive: this.state.currentTab === "configuration",
            },
          ]}
        />

        {this.state.currentTab === "runHistory" ? (
          <JobRunHistoryTable job={job} />
        ) : this.state.currentTab === "configuration" ? (
          <JobConfiguration job={job} />
        ) : null}

        <JobFormModal
          isEdit={true}
          job={JSON.parse(job.json)}
          isOpen={this.state.currentDialog === "edit"}
          closeModal={closeDialog}
        />
        <JobDelete
          jobId={params.id}
          open={this.state.currentDialog === "destroy"}
          onSuccess={jobDeleteSuccess}
          onClose={closeDialog}
        />
      </Page>
    );
  }
}

export default JobDetailPage;
