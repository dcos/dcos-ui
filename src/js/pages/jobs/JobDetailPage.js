import classNames from "classnames";
import prettycron from "prettycron";
/* eslint-disable no-unused-vars */
import React from "react";
/* eslint-enable no-unused-vars */
import { routerShape } from "react-router";
import mixin from "reactjs-mixin";

import JobCreateEditFormModal from "#PLUGINS/jobs/src/js/JobCreateEditFormModal";
import TaskStates from "#PLUGINS/services/src/js/constants/TaskStates";
import jobsMenu from "#PLUGINS/jobs/src/js/jobsMenu";
import JobDelete from "#PLUGINS/jobs/src/js/JobDelete";
import JobsBreadcrumbs from "../../components/JobsBreadcrumbs";

import Icon from "../../components/Icon";
import Page from "../../components/Page";
import TimeAgo from "../../components/TimeAgo";
import TabsMixin from "../../mixins/TabsMixin";
import StringUtil from "../../utils/StringUtil";
import JobConfiguration from "./JobConfiguration";
import { DIALOGS } from "./JobDetailPageContainer";
import JobRunHistoryTable from "./JobRunHistoryTable";

class JobDetailPage extends mixin(TabsMixin) {
  constructor() {
    super(...arguments);

    this.tabs_tabs = {
      runHistory: "Run History",
      configuration: "Configuration"
    };

    this.state = {
      currentTab: Object.keys(this.tabs_tabs).shift()
    };
  }

  getNavigationTabs() {
    return <ul className="menu-tabbed">{this.tabs_getUnroutedTabs()}</ul>;
  }

  getPrettySchedule(job) {
    const schedules = job.getSchedules();
    if (schedules == null || schedules.length === 0) {
      return null;
    }

    const schedule = schedules[0];
    if (schedule.enabled) {
      return prettycron.toString(schedule.cron);
    }
  }

  getSubTitle(job) {
    const nodes = [];
    const scheduleText = this.getPrettySchedule(job);
    let longestRunningTask = null;
    const longestRunningActiveRun = job
      .getActiveRuns()
      .getLongestRunningActiveRun();

    if (longestRunningActiveRun != null) {
      longestRunningTask = longestRunningActiveRun
        .getTasks()
        .getLongestRunningTask();
    }

    if (longestRunningTask == null && scheduleText == null) {
      return null;
    }

    if (scheduleText != null) {
      nodes.push(
        <p className="text-overflow flush" key="schedule-text">
          <Icon
            className="icon-margin-right"
            key="schedule-icon"
            color="grey"
            id="repeat"
            size="mini"
          />
          <span>Scheduled {StringUtil.lowercase(scheduleText)}</span>
        </p>
      );
    }

    if (longestRunningTask != null) {
      const dateCompleted = longestRunningTask.getDateCompleted();

      const status = TaskStates[longestRunningTask.getStatus()];
      const statusClasses = classNames("job-details-header-status", {
        "text-success":
          status.stateTypes.includes("success") &&
          !status.stateTypes.includes("failure"),
        "text-danger": status.stateTypes.includes("failure")
      });

      let timePrefix = null;
      const shouldSuppressRelativeTime = dateCompleted == null;

      if (shouldSuppressRelativeTime) {
        timePrefix = "for ";
      }

      nodes.push(
        <span className={statusClasses} key="status-text">
          {status.displayName}
        </span>,
        <span key="time-running">
          <TimeAgo
            prefix={timePrefix}
            suppressSuffix={shouldSuppressRelativeTime}
            time={longestRunningActiveRun.getDateCreated()}
          />
        </span>
      );
    }

    return nodes;
  }

  renderConfigurationTabView(job) {
    return <JobConfiguration job={job} />;
  }

  renderRunHistoryTabView(job) {
    return <JobRunHistoryTable job={job} />;
  }

  getActions() {
    const job = this.props.job;

    const customActionHandlers = {
      edit: this.props.handleEditButtonClick,
      delete: this.props.handleDestroyButtonClick
    };

    return jobsMenu(job, customActionHandlers);
  }

  getTabs() {
    const activeTab = this.state.currentTab;

    return [
      {
        label: "Run History",
        callback: () => {
          this.setState({ currentTab: "runHistory" });
        },
        isActive: activeTab === "runHistory"
      },
      {
        label: "Configuration",
        callback: () => {
          this.setState({ currentTab: "configuration" });
        },
        isActive: activeTab === "configuration"
      }
    ];
  }

  render() {
    // TaskDetailView
    if (this.props.params.taskID) {
      return this.props.children;
    }

    const { params, job, jobTree } = this.props;

    return (
      <Page>
        <Page.Header
          actions={this.getActions()}
          breadcrumbs={<JobsBreadcrumbs tree={jobTree} item={job} />}
          tabs={this.getTabs()}
        />
        {this.tabs_getTabView(job)}
        <JobCreateEditFormModal
          isEdit={true}
          job={job}
          open={this.props.jobActionDialog === DIALOGS.EDIT}
          onClose={this.props.closeDialog}
        />
        <JobDelete
          jobId={params.id}
          open={this.props.jobActionDialog === DIALOGS.DESTROY}
          onSuccess={this.props.onJobDeleteSuccess}
          onClose={this.props.closeDialog}
        />
      </Page>
    );
  }
}

JobDetailPage.contextTypes = {
  router: routerShape
};

module.exports = JobDetailPage;
