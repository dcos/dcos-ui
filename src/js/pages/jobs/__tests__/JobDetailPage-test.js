import React from "react";
import renderer from "react-test-renderer";

import JobDetailPage from "../JobDetailPage";

describe("JobDetailPage", () => {
  describe("#renderBreadcrumbStates", () => {
    const testCases = [
      {
        name: "with no schedule",
        schedules: [],
        jobTask: null
      },
      {
        name: "with enabled schedules",
        schedules: [
          {
            enabled: true,
            cron: "0 0 20 * *"
          }
        ],
        jobTask: null
      },
      {
        name: "with job runs",
        schedules: [],
        jobTask: { status: "TASK_RUNNING" }
      },
      {
        name: "with job task and schedule",
        schedules: [
          {
            enabled: true,
            cron: "0 0 20 * *"
          }
        ],
        jobTask: { status: "TASK_RUNNING" }
      }
    ];

    for (const testCase of testCases) {
      it(`renders ${testCase.name}`, () => {
        const item = {
          schedules: {
            nodes: testCase.schedules
          },
          jobRuns: {
            longestRunningActiveRun: {
              tasks: {
                longestRunningTask: testCase.jobTask
              }
            }
          }
        };
        expect(
          renderer
            .create(
              <div>{JobDetailPage.prototype.renderBreadcrumbStates(item)}</div>
            )
            .toJSON()
        ).toMatchSnapshot();
      });
    }
  });
});
