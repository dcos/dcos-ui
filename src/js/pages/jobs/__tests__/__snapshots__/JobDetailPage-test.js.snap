// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobDetailPage #renderBreadcrumbStates renders with enabled schedules 1`] = `
<div>
  <div
    className="breadcrumb__content breadcrumb__content--supplemental"
  >
    <div
      className="tooltip-wrapper text-align-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <svg
        className="icon icon-grey icon-mini"
      >
        <use
          xlinkHref="#icon-system--repeat"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`JobDetailPage #renderBreadcrumbStates renders with job runs 1`] = `
<div>
  <div
    className="breadcrumb__content breadcrumb__content--supplemental"
  >
    <div
      className="service-page-header-status muted"
    >
      (
      Running
      )
    </div>
  </div>
</div>
`;

exports[`JobDetailPage #renderBreadcrumbStates renders with job task and schedule 1`] = `
<div>
  <div
    className="breadcrumb__content breadcrumb__content--supplemental"
  >
    <div
      className="tooltip-wrapper text-align-center"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <svg
        className="icon icon-grey icon-mini"
      >
        <use
          xlinkHref="#icon-system--repeat"
        />
      </svg>
    </div>
  </div>
  <div
    className="breadcrumb__content breadcrumb__content--supplemental"
  >
    <div
      className="service-page-header-status muted"
    >
      (
      Running
      )
    </div>
  </div>
</div>
`;

exports[`JobDetailPage #renderBreadcrumbStates renders with no schedule 1`] = `<div />`;
